nvflare/app_opt/xgboost/histogram_based_v2/sec/client_handler.py

```diff
@@ -291,11 +354,31 @@ def _process_after_all_gather_v(self, fl_ctx: FLContext):
         result = self.data_converter.encode_aggregation_result(final_result, fl_ctx)
         fl_ctx.set_prop(key=Constant.PARAM_KEY_RCV_BUF, value=result, private=True, sticky=False)
 
+    def _process_after_all_gather_v_horizontal(self, fl_ctx: FLContext):
+        encrypted_histograms = fl_ctx.get_prop(Constant.PARAM_KEY_RCV_BUF)
+        rank = fl_ctx.get_prop(Constant.PARAM_KEY_RANK)
+        if not isinstance(encrypted_histograms, CKKSVector):
+            return self._abort(f"rank {rank}: expect a CKKSVector but got {type(encrypted_histograms)}", fl_ctx)
+
+        histograms = encrypted_histograms.decrypt(secret_key=self.tenseal_context.secret_key())
+        result = self.data_converter.encode_histograms_result(histograms, fl_ctx)
+        fl_ctx.set_prop(key=Constant.PARAM_KEY_RCV_BUF, value=result, private=True, sticky=False)
+
     def handle_event(self, event_type: str, fl_ctx: FLContext):
         if event_type == EventType.START_RUN:
             self.public_key, self.private_key = generate_keys(self.key_length)
             self.encryptor = Encryptor(self.public_key, self.num_workers)
             self.decrypter = Decrypter(self.private_key, self.num_workers)
             self.adder = Adder(self.num_workers)
+            try:
+                if tenseal_imported:
+                    self.tenseal_context = load_tenseal_context_from_workspace(self.tenseal_context_file, fl_ctx)
+                else:
+                    self.debug(fl_ctx, "Tenseal module not loaded, horizontal secure XGBoost is not supported")
```

Comment:
Maybe we should warn them in the `__init__` method if tenseal is not installed?

Reply:
If only vertical is used, this seems to be excessive warnings.